### DISCUSSION TITLE
AWS directory structure

### DISCUSSION THREAD

**Discussion by mklatsky** [terramate-io/terramate#545] [Q&A]
I am starting to try terramate and am unsure of the proper directory structure to keep my code as DRY as possible. Currently, my structure looks like this:

```
project-
├── backend.tf.vars
├── globals.tm.hcl
├── modules
├── provider.tf
├── stacks
│   ├── globals.tm.hcl
│   └── sandbox
│       ├── api_gateway
│       │   ├── api_gateway.tf
│       │   ├── backend.tf
│       │   └── config.tm.hcl
|      ├── cloudfront
│       │   ├── backend.tf
│       │   ├── cloudfront.tf
│       │   └── config.tm.hcl
```

I have several AWS accounts- sandbox, testing and prod. Within each account, I might need to launch multiple "projects". Currently, I have a hierarchy in my stacks showing an account (sandbox), then the individual services I am launching. This seems a bit more complex than I expected. 

Any insight or examples would be most appreciated.

**Comment by sergeyshevch** [comment]
I have the same structure for all accounts. Also, I have one more level for region (and global resources) separately. 
If you maximally reuse modules it will reduce code that you need to duplicate across accounts

**Comment by soerenmartius** [comment]
Hi @mklatsky, and thanks for your question. We are currently working on some best practices regarding structures in AWS multi-account and GC multi-project projects and will publish more content on this soon. In the meantime, let me sync on your specific request with the team. I will come back to you shortly.

**Reply by b100to** [reply]
When can I expect to see it? It's not easy because of directory structure concerns.

**Comment by mariux** [comment]
Hi @mklatsky 

> This seems a bit more complex than I expected. 
> Any insight or examples would be most appreciated.

To be able to answer your question: What was you expectation and where do you see the complexity? 

In plain Terraform set ups, where you just want to orchestrate the execution of Terraform over multiple stacks, you need to create a stack config e.g. by using `terramate create` on each stack.

Once you enter the space of code generation you can eliminate a lot of duplicate code you need to maintain manually by generating it. The most useful start here would to automate generation of backend and provider configurations and the terraform version to use. We recommend to always pin all versions to a specific one: Terraform, Terraform Providers and Terraform Modules versions.

Please let us know what the unexpected complexity for you is and i am happy to try to give some examples guiding you into how we might solve this.
